### PR TITLE
Add deletes to the range tree.

### PR DESCRIPTION
**Pull request by BramGruneir**
This finishes the work on the red black range tree by adding in range deletion and attaching it to the adminMerge command.

### CODE REVIEW

File: storage/client_range_tree_test.go [hunk 1]
```diff
@@ -211,9 +193,9 @@ func TestSetupRangeTree(t *testing.T) {
 	VerifyTree(t, tree, nodes, "setup")
 }
 
-// TestInsert is a similar to the one in range_tree_test but this one performs
-// actual splits.
-func TestInsert(t *testing.T) {
+// TestTree is a similar to the one in range_tree_test but this one performs
```
Comment (tamird) [inline storage/client_range_tree_test.go]: a similar what?

File: storage/client_range_tree_test.go [hunk 2]
```diff
@@ -226,11 +208,22 @@ func TestInsert(t *testing.T) {
...
-	for _, key := range keys {
+	for _, key := range keysInsert {
 		if err := db.AdminSplit(key); err != nil {
 			t.Fatal(err)
 		}
 		tree, nodes := loadTree(t, db)
 		VerifyTree(t, tree, nodes, key)
 	}
+
+	// To test merging, we just call merged on the lowest key to merge all
```
Comment (tamird) [inline storage/client_range_tree_test.go]: s/merged/merge/?

File: storage/range_tree.go [hunk 3]
```diff
@@ -443,3 +454,407 @@ func (tc *treeContext) insertCase5(node *proto.RangeTreeNode) error {
...
+		return err
+	}
+	node, err := tc.getNode(key)
+	if err != nil {
+		return err
+	}
+	if node == nil {
+		return util.Errorf("could not find range tree node with the key %s", key)
+	}
+	err = tc.delete(node)
```
Comment (tamird) [inline storage/range_tree.go]: nit: `err := ..;err != nil {`

File: storage/range_tree.go [hunk 4]
```diff
@@ -443,3 +454,407 @@ func (tc *treeContext) insertCase5(node *proto.RangeTreeNode) error {
...
+	} else if node.RightKey != nil {
+		if child, err = tc.getNode(node.RightKey); err != nil {
+			return err
+		}
+	}
+	if !isRed(node) {
+		// Paint the node to the color of the child node.
+		node.Black = !isRed(child)
+		tc.setNode(node)
+		if err = tc.deleteCase1(node); err != nil {
```
Comment (tamird) [inline storage/range_tree.go]: why not `:=` here?

File: storage/range_tree.go [hunk 5]
```diff
@@ -443,3 +454,407 @@ func (tc *treeContext) insertCase5(node *proto.RangeTreeNode) error {
...
+			return err
+		}
+	}
+	_, err = tc.replaceNode(node, child)
```
Comment (tamird) [inline storage/range_tree.go]: this `err` looks unchecked.

File: storage/range_tree.go [hunk 6]
```diff
@@ -443,3 +454,407 @@ func (tc *treeContext) insertCase5(node *proto.RangeTreeNode) error {
...
+		parent, err := tc.getNode(node.ParentKey)
+		if err != nil {
+			return err
+		}
+		parent.Black = false
+		sibling.Black = true
+		tc.setNode(parent)
+		tc.setNode(sibling)
+		if parent.LeftKey.Equal(node.Key) {
+			if _, err = tc.rotateLeft(parent); err != nil {
```
Comment (tamird) [inline storage/range_tree.go]: `:=`

File: storage/range_tree.go [hunk 7]
```diff
@@ -443,3 +454,407 @@ func (tc *treeContext) insertCase5(node *proto.RangeTreeNode) error {
...
+				return err
+			}
+		} else {
+			if _, err = tc.rotateRight(parent); err != nil {
```
Comment (tamird) [inline storage/range_tree.go]: `:=`

File: storage/client_range_tree_test.go (hunk 1 above)
Comment (BramGruneir) [inline storage/client_range_tree_test.go]: fixed.

File: storage/client_range_tree_test.go (hunk 2 above)
Comment (BramGruneir) [inline storage/client_range_tree_test.go]: hmm.  I know I wrote AdminMerge on this.  I maybe reverted it accidentally.  Anyway, fixed.

File: storage/range_tree.go (hunk 3 above)
Comment (BramGruneir) [inline storage/range_tree.go]: So where did we end up with on shadowing err?  Do we allow it or not?  I've been just avoiding shadowing everywhere.

Comment (BramGruneir) [inline storage/range_tree.go]: done.

File: storage/range_tree.go (hunk 4 above)
Comment (BramGruneir) [inline storage/range_tree.go]: done

File: storage/range_tree.go (hunk 5 above)
Comment (BramGruneir) [inline storage/range_tree.go]: good catch, fixed

File: storage/range_tree.go (hunk 6 above)
Comment (BramGruneir) [inline storage/range_tree.go]: fixed

File: storage/range_tree.go (hunk 7 above)
Comment (BramGruneir) [inline storage/range_tree.go]: fixed

File: storage/range_tree.go (hunk 3 above)
Comment (tamird) [inline storage/range_tree.go]: the TL;DR is that it's preferable to shadow in all cases where the shadowing is unproblematic. that is, if you can avoid ever writing `err =` for a particular scope, you should. If you can't avoid that assignment then you need to be more deliberate about shadowing vs assignment. this is probably easier to describe deeply in person.

Comment (BramGruneir) [inline storage/range_tree.go]: So I actually went through the whole file and fixed all of these errors to shadow and return in a single if statement.

Comment (BramGruneir) [inline storage/range_tree.go]: Perhaps we should have a page in the wiki about our agreed upon code style.  Especially when it comes to things like this, where examples are probably needed.

**Comment by BramGruneir** [comment]
PTAL
All comments addressed, I think this is ready to merge.

**Comment by BramGruneir** [comment]
@tamird can you take another look when you have a min?

File: storage/range_tree_test.go [hunk 8]
```diff
@@ -846,4 +1465,31 @@ func TestInsert(t *testing.T) {
...
+	// was chosen experimentally.
+	keysDelete := []string{"f", "e", "d", "c", "x",
+		"g", "h", "i", "j", "k", "l",
+		"z", "n", "o", "p", "q",
+		"t", "u", "v", "w", "y",
+		"r", "b", "a", "s"}
+
+	for _, key := range keysDelete {
+		node, err := tc.getNode(proto.Key(key))
+		if err != nil {
```
Comment (tamird) [inline storage/range_tree_test.go]: scoping here as discussed

File: storage/range_tree.go [hunk 9]
```diff
@@ -431,15 +440,418 @@ func (tc *treeContext) insertCase5(node *proto.RangeTreeNode) error {
...
+	if node.LeftKey != nil && node.RightKey != nil {
+		left, err := tc.getNode(node.LeftKey)
+		if err != nil {
+			return err
+		}
+		predecessor, err := tc.getMaxNode(left)
+		if err != nil {
+			return err
+		}
+		node, predecessor, err = tc.swapNodes(node, predecessor)
```
Comment (tamird) [inline storage/range_tree.go]: s/prodecessor/_/

File: storage/range_tree_test.go [hunk 10]
```diff
@@ -846,4 +1465,31 @@ func TestInsert(t *testing.T) {
...
+			t.Fatal(err)
+		}
+		if err = tc.delete(node); err != nil {
```
Comment (tamird) [inline storage/range_tree_test.go]: =

**Comment by tamird** [comment]
LG

File: storage/range_tree_test.go (hunk 8 above)
Comment (BramGruneir) [inline storage/range_tree_test.go]: done.

File: storage/range_tree.go (hunk 9 above)
Comment (BramGruneir) [inline storage/range_tree.go]: done

File: storage/range_tree_test.go (hunk 10 above)
Comment (BramGruneir) [inline storage/range_tree_test.go]: done

**Pull request merged**